Source/PeleC.H

```diff
@@ -125,45 +129,82 @@ public:
   // Initialize grid data at problem start-up.
   void initData() override;
 
-#ifdef PELEC_SPRAY
+#ifdef PELEC_USE_SPRAY
 
-  // Read particle-related inputs
-  static void readParticleParams();
+  // Read spray related inputs
+  static void readSprayParams();
 
   // Define particle related variables
   static void defineParticles();
 
   // Initialize particle locations and velocities (and strengths if relevant)
   void initParticles();
 
-  // Timestamp particles
-  void particleTimestamp(int ngrow);
-
-  // Default verbosity of Particle class
-  static int particle_verbose;
+  // Create relevant particle data
+  void createDataParticles();
 
   // How to initialize at restart
-  void particlePostRestart(
-    const std::string& restart_file, bool is_checkpoint = true);
+  void postRestartParticles(bool is_checkpoint = true);
+
+  // Post timestep particle functions
+  void postTimeStepParticles(int iteration);
+
+  // Post init for spray particles
+  void postInitParticles();
 
   // Redistribute
-  virtual void particleRedistribute(
-    int lbase = 0, int nGrow = 0, int local = 0, bool init_part = false);
+  void particle_redistribute(int lbase = 0, bool init_part = false) override;
+
+  // Setup ghost and virtual particles and do moveKickDrift
+  void particleMKD(
+    amrex::Real time,
+    amrex::Real dt,
+    int sub_iteration,
+    int sub_ncycle,
+    int amr_ncycle);
+
+  // Do the moveKick for active and ghost particles
+  void particleMK(
+    amrex::Real time,
+    amrex::Real dt,
+    int sub_iteration,
+    int sub_ncycle,
+    int amr_ncycle);
 
   // Setup virtual particles if necessary
-  void setupVirtualParticles();
+  static void setupVirtualParticles(const int level, const int finest_level);
 
   // Remove virtual particles if necessary
-  void removeVirtualParticles();
+  static void removeVirtualParticles(const int level);
 
   // Setup ghost particles (for finer levels) if necessary
-  void setupGhostParticles(int ngrow);
+  static void
+  setupGhostParticles(const int level, const int finest_level, const int ngrow);
 
   // Remove ghost particles (for this level) if necessary
-  void removeGhostParticles();
+  static void removeGhostParticles(const int level);
 
   // Time step control based on particles
-  void particleEstTimeStep(amrex::Real& est_dt);
+  void estTimeStepParticles(amrex::Real& est_dt);
+
+  // Initialize the temporary spray source
+  void defineSpraySource(int amr_ncycle);
+  amrex::MultiFab tmp_spray_source;
```

Comment:
Is this necessary or can it be rolled into `spray_source`?

Reply:
It could be possible to use `spray_source` but I would have to change the number of ghost cells in `spray_source`. There are times when particles might use more ghost cells in the source MF than would actually be needed for the state update. Also, the temporary source MF only has the components it needs.

Reply:
ok sounds reasonable to keep as is.